core/src/main/java/org/apache/calcite/rex/RexCall.java

```diff
@@ -80,7 +80,6 @@ protected RexCall(
     this.nodeCount = RexUtil.nodeCount(1, this.operands);
     assert op.getKind() != null : op;
     assert op.validRexOperands(operands.size(), Litmus.THROW) : this;
-    assert op.kind != SqlKind.IN || this instanceof RexSubQuery;
```

Comment:
Just in case, this assertion was added in `CALCITE-4173`, however, it breaks existing applications that might create their own IN RexCalls.

Reply:
Agree, we should remove the assertion.

Reply:
No, this assertion is needed. We do not support IN in RexCall.

Reply:
@julianhyde IMHO the assertion is a drastic breaking change. As @vlsi says, what happens with existing applications that may create their own IN calls (that then are implemented by their own mechanisms / conventions)? Should this be suddenly unsupported in a minor version upgrade?

Reply:
I address this in the JIRA case. We never officially supported IN in RexCall. We allowed it for limited cases only (e.g. RelToSql).

Calcite algebra isn't a free-for-all, where you can add your own built-in operator and then holler when it is de-supported. If it was, people would end up choking on each other's garbage.

This isn't 'drastic'. Most people aren't using this.

This isn't 'sudden'. For years I've been arguing in favor of expanding SQL IN lists to ORs.

Reply:
I'm +1 to Julian's idea.

Reply:
> This isn't 'sudden'. For years I've been arguing in favor of expanding SQL IN lists to ORs.

One thing is expanding IN lists to ORs (which I think is totally fine, and Calcite should continue doing it), and a different thing is forbidding creating an IN list call (which is the "new" change introduced by CALCITE-4173).

> This isn't 'drastic'. Most people aren't using this.

I do not know about "most people", I know that my application is using it. In my specific example, even though IN lists are expanded as ORs, under certain circumstances my application re-contracts them again as an IN list. The reason for that is allowing to execute later an IndexScan exploiting some Lucene indices, where the IN list call can be implemented as a Lucene's TermInSetQuery.
Other applications can have other use cases, but I would say that having the possibly to define an IN list call in relational algebra is something useful that should not be forbidden.

Reply:
> having the possibly to define an IN list call in relational algebra (even though Calcite core does not do it) is something useful that should not be forbidden.

I agree. This feature is precisely that facility.